### PR TITLE
Support custom rule sets for Ktlint

### LINKED ISSUE #1220: Allow enabling custom rulesets for ktlint
Follow up from https://github.com/diffplug/spotless/issues/409#issuecomment-1107055624

I had been watching the issue around this before, but it was closed after just the first request (experimental ruleset) was completed. Currently my team runs a second formatting plugin (https://github.com/JLLeitschuh/ktlint-gradle) that reruns ktlint so that we can use a custom ruleset that helps us avoid some common pitfalls in our codebase.

The way that we use this today looks roughly like:

build.gradle.kts
```
plugins {
    java
    kotlin("jvm")
    id("com.diffplug.spotless")
    // The extra ktlint plugin
    id("org.jlleitschuh.gradle.ktlint")
}

dependencies {
    // Add the ruleset from a submodule
    ktlintRuleset(project(":linting"))
}

spotless {
    kotlin {
        target("**/*.kt")
        ktlint()
    }
}

// After running spotlessApply or spotlessCheck, run a last check with the custom rules
tasks.spotlessKotlin {
    finalizedBy(
        "ktlintCheck"
    )
}
```

### PR DESCRIPTION
**Pull request by Goooler**
Closes #1220.

### CODE REVIEW

File: plugin-gradle/src/test/java/com/diffplug/gradle/spotless/KotlinExtensionTest.java [hunk 1]
```diff
@@ -141,6 +141,30 @@ void testSetEditorConfigCanOverrideEditorConfigFile() throws IOException {
...
+				"            \"io.nlopez.compose.rules:ktlint:0.3.3\"",
+				"        ])",
+				"        .editorConfigOverride([",
+				"            ktlint_function_naming_ignore_when_annotated_with: \"Composable\",",
+				"        ])",
+				"    }",
+				"}");
+		setFile("src/main/kotlin/Main.kt").toResource("kotlin/ktlint/listScreen.dirty");
+		String buildOutput = gradleRunner().withArguments("spotlessCheck").buildAndFail().getOutput();
+		assertThat(buildOutput).contains("Composable functions that return Unit should start with an uppercase letter.");
```
Comment (Goooler) [inline plugin-gradle/src/test/java/com/diffplug/gradle/spotless/KotlinExtensionTest.java:165]: https://mrmans0n.github.io/compose-rules/rules/#naming-composable-functions-properly

File: plugin-maven/README.md [hunk 2]
```diff
@@ -416,6 +416,9 @@ Additionally, `editorConfigOverride` options will override what's supplied in `.
     <ij_kotlin_allow_trailing_comma>true</ij_kotlin_allow_trailing_comma>
     <ij_kotlin_allow_trailing_comma_on_call_site>true</ij_kotlin_allow_trailing_comma_on_call_site>
   </editorConfigOverride>
+  <customRuleSets> <!-- optional -->
+    <value>io.nlopez.compose.rules:ktlint:0.3.3</value>
```
Comment (Goooler) [inline plugin-maven/README.md:420]: Should we use `value`, `String`, or something else as the node key?

**Comment by Goooler** [comment]
May merge #1897 first.

File: plugin-maven/README.md (hunk 2 above)
Comment (nedtwigg) [reply plugin-maven/README.md:420]: I think `value` is good, but I'm happy to use anything else if you prefer.

**Comment by nedtwigg** [comment]
Looks great! Fixup merge conflicts and I'll merge. Anything else you'd like to work on for the next release, or should I publish after this and 

- #1881 

get merged?

**Comment by Goooler** [comment]
Yeah, we can publish a new version after these 2 get merged.

File: plugin-maven/src/main/java/com/diffplug/spotless/maven/kotlin/Ktlint.java [hunk 3]
```diff
@@ -46,7 +50,10 @@ public FormatterStep newFormatterStep(final FormatterStepConfig stepConfig) {
 		if (editorConfigOverride == null) {
 			editorConfigOverride = new HashMap<>();
 		}
+		if (customRuleSets == null) {
+			customRuleSets = Collections.emptyList();
+		}
```
Comment (Goooler) [inline plugin-maven/src/main/java/com/diffplug/spotless/maven/kotlin/Ktlint.java:55]: Seems these reassigning is unnecessary?

Comment (Goooler) [reply plugin-maven/src/main/java/com/diffplug/spotless/maven/kotlin/Ktlint.java:55]: They are necessary, `TreeMap` and `HashSet` need `NonNull` inputs.

**Comment by nedtwigg** [comment]
Released in `plugin-gradle 6.23.0` and `plugin-maven 2.41.0`.

**Pull request merged**